### PR TITLE
Fix event subscription/unsubscription with null handler

### LINKED ISSUE #2033: Adding null-valued `EventHandler`s to a fake leads to `UserCallbackException`
Hi!

I hope the title is already telling enough, should be pretty easy to reproduce.

Example:
```csharp
public interface ITestInterface
{
    public event EventHandler? TestEvent;
}

...

public void TestExample()
{
    var fake = A.Fake<ITestInterface>();
    fake.TestEvent += null;
}
```

Executing `TestExample` throws the following exception:
```
... threw exception. FakeItEasy.UserCallbackException: Call filter <Subscription to any event of Faked TestExample> threw an exception. See inner exception for details. ---> System.ArgumentNullException: Value cannot be null. (Parameter 'key').
    at System.ThrowHelper.ThrowArgumentNullException(String name)
   at System.Collections.Concurrent.ConcurrentDictionary`2.TryGetValue(TKey key, TValue& value)
   at System.Collections.Concurrent.ConcurrentDictionary`2.ContainsKey(TKey key)
   at FakeItEasy.Core.EventHandlerArgumentProviderMap.HasArgumentProvider(Delegate eventHandler) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Core\EventHandlerArgumentProviderMap.cs:line 23
   at FakeItEasy.Core.EventCall.IsEventRaiser() in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Core\EventCall.cs:line 56
   at FakeItEasy.EventAction.AddAnyEventAction.Matches(EventCall eventCall) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\EventAction.cs:line 51
   at FakeItEasy.EventAction.Matches(IFakeObjectCall call) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\EventAction.cs:line 41
   at FakeItEasy.Configuration.BuildableCallRule.WherePredicate.Matches(IFakeObjectCall call) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Configuration\BuildableCallRule.cs:line 270
--- End of inner exception stack trace ---
    at FakeItEasy.Configuration.BuildableCallRule.WherePredicate.Matches(IFakeObjectCall call) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Configuration\BuildableCallRule.cs:line 274
   at FakeItEasy.Configuration.BuildableCallRule.<>c__DisplayClass30_0.<IsApplicableTo>b__0(WherePredicate x) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Configuration\BuildableCallRule.cs:line 127
   at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)
   at FakeItEasy.Configuration.BuildableCallRule.IsApplicableTo(IFakeObjectCall fakeObjectCall) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Configuration\BuildableCallRule.cs:line 127
   at FakeItEasy.Core.FakeManager.ApplyBestRule(IInterceptedFakeObjectCall fakeObjectCall) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Core\FakeManager.cs:line 247
   at FakeItEasy.Core.FakeManager.FakeItEasy.Core.IFakeCallProcessor.Process(InterceptedFakeObjectCall fakeObjectCall) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Core\FakeManager.cs:line 178
   at FakeItEasy.Creation.CastleDynamicProxy.CastleDynamicProxyGenerator.ProxyInterceptor.Intercept(IInvocation invocation) in D:\a\FakeItEasy\FakeItEasy\src\FakeItEasy\Creation\CastleDynamicProxy\CastleDynamicProxyGenerator.cs:line 186
   at Castle.DynamicProxy.AbstractInvocation.Proceed()
   at Castle.Proxies.ObjectProxy.add_TestEvent(EventHandler value)
   at ...
```

Thanks for any help!

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fixes #2033

### CODE REVIEW

File: tests/FakeItEasy.Specs/EventRaisingSpecs.cs [hunk 1]
```diff
@@ -526,5 +526,31 @@ public void RaisingInternalEvent(ClassWithInternalEventVisibleToDynamicProxy fak
...
+        {
+            "Given a fake with an event"
+                .x(() => fake = A.Fake<IEvents>());
+
+            "When unsubscribing from the event with a null handler"
+                .x(() => exception = Record.Exception(() => fake.UnsubscribedEvent -= null));
+
+            "Then it should not throw"
+                .x(() => exception.Should().BeNull());
+        }
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/EventRaisingSpecs.cs:554]: I don't love adding these specs to `EventRaisingSpecs`, which is about _raising_ events, not subscribing or unsubscribing. But we don't have specs for event subscription/unsubscription, and I also don't love adding a new spec file just for that.

Arguably, `EventRaisingSpecs` covers not only event raising, but also subscription/unsubscription. So maybe we should just rename it to `EventSpecs` or something like that.

Comment (blairconrad) [reply tests/FakeItEasy.Specs/EventRaisingSpecs.cs:554]: I agree. Renaming to `EventSpecs` suits me. We can do here or later, as pleases you.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. Your changes look sufficient to me. But they were bigger than I anticipated. I may be missing something, but https://github.com/blairconrad/FakeItEasy/commit/c0091b107501f9e92164169091191f1c01b3bc05 also makes the specs pass with less disruption to the original code (and less mucking about with nullable hints). Thoughts?

**Comment by thomaslevesque** [comment]
I like your approach! It's simpler, and doesn't require `EventCall.EventHandler` to be nullable.
Do you mind if I just cherry-pick your commit to replace my changes?

**Comment by blairconrad** [comment]
> Do you mind if I just cherry-pick your commit to replace my changes?

I do. I'm very protective of my IP.

Don't be silly.

**Review by blairconrad — APPROVED** [review]
Thanks, @thomaslevesque!

**Comment by thomaslevesque** [comment]
Thanks for the assist!

**Comment by blairconrad** [comment]
teamwork

**Pull request merged**